api/envoy/type/v3/ratelimit_strategy.proto

```diff
@@ -0,0 +1,85 @@
+syntax = "proto3";
+
+package envoy.type.v3;
+
+import "envoy/type/v3/ratelimit_unit.proto";
+import "envoy/type/v3/token_bucket.proto";
+
+import "xds/annotations/v3/status.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.type.v3";
+option java_outer_classname = "RatelimitStrategyProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/type/v3;typev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#not-implemented-hide:]
+// [#protodoc-title: Rate Limit Strategies]
+
+message RateLimitStrategy {
+  // Choose between allow all and deny all.
+  enum BlanketRule {
+    ALLOW_ALL = 0;
+    DENY_ALL = 1;
+  }
+
+  // Limit the number of requests per second.
+  message RequestsPerTimeUnit {
+    // The number of requests per unit of time to allow. Requests over the limit
+    // are denied.
+    // If set to 0, deny all (equivalent to BlanketRule.DENY_ALL).
+    uint64 requests_per_time_unit = 1;
+
+    // The unit of time. Ignored when requests_per_time_unit is 0 (deny all).
+    RateLimitUnit time_unit = 2;
+  }
+
+  oneof strategy {
+    option (validate.required) = true;
+
+    // Allow or Deny the requests.
+    // If unset, allow all.
+    BlanketRule blanket_rule = 1;
+
+    // A number of requests per time unit, f.e. requests per second.
+    RequestsPerTimeUnit requests_per_time_unit = 2;
+
+    // Limit the requests by consuming tokens from the Token Bucket.
+    // Allow the same number of requests as the number of tokens available in
+    // the token bucket.
+    TokenBucket token_bucket = 3;
+  }
+}
+
+// Queue all requests. Useful as an intermediate strategy while loading actual
+// rate limiting assignment asynchronously (f.e. from a remote service).
+// To avoid memory exhaust, queue configuration allows to
+// to impose constraints on the maximum number of pending requests
+// in the queue.
+message RateLimitStrategyQueue {
```

Comment:
I do not think we need this for Envoy. I'm unsure how the queue length can be estimated with any sort of precision. Envoy has overload manager that is used to avoid OOMs. Overload manager can dynamically lower idle timeouts under memory pressure, causing requests that were waiting for the assignment to be reset and free resources.

Is this something that gRPC would use?

Reply:
Thanks for pointing this out. I need a bit more info to get a better understanding of how this works. Let's talk IRL about this manager.

Reply:
After discussing gRPC implementation details with Eric, I think it's fair to keep the logic for blocking requests under the hood - and let the implementers choose the best approach. I'll remove `RateLimitStrategyQueue`.

Reply:
Removed in https://github.com/envoyproxy/envoy/pull/19793/commits/0e3eaf3b634e52f6966a7e32ef202b2d6fbc42e6. Now it's captured in the filter's `BlockingBehavior`. Also marked as not implemented.